File: Leetcode/sort_and_search/154_FindMinimumInRotatedSortedArrayII/Solution.java
class Solution {
    public int findMin(int[] nums) {
        return search(nums, 0, nums.length-1);
    }
    private int search(int[] nums, int start, int end){
        if(end-start==0)
            return nums[start];
        if(end-start==1)
            return Math.min(nums[start], nums[end]);
        int mid = start + (end - start)/2;
        if(mid-1>=start && nums[mid]<nums[mid-1])
            return nums[mid];
        if(nums[mid]==nums[end])
            return search(nums, start, end-1);
        if(nums[start]<=nums[mid]){
            if(nums[start]<=nums[mid] && nums[mid]<=nums[end]){
                return search(nums, start, mid-1);
            }else{
                return search(nums, mid+1, end);
            }
        }
        return search(nums, start, mid-1);
    }
}

File: Leetcode/sort_and_search/154_FindMinimumInRotatedSortedArrayII/Solution.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==1:
            return nums[0]
        if len(nums)==2:
            return min(nums[0], nums[1])
        start, end = 0, len(nums)-1
        mid = start + (end-start)//2
        if mid-1>=0 and nums[mid]<nums[mid-1]:
            return nums[mid]
        if nums[mid]==nums[end]:
            return self.findMin(nums[:end])
        if nums[start]<=nums[mid]:
            if nums[start]<=nums[mid]<=nums[end]:
                return self.findMin(nums[:mid])
            else:
                return self.findMin(nums[mid+1:])
        if nums[mid]<=nums[end]:
            return self.findMin(nums[:mid])
